runtime/compiler/runtime/RelocationRuntime.cpp

```diff
@@ -1435,3 +1427,33 @@ TR_JITaaSRelocationRuntime::allocateSpaceInDataCache(uintptr_t metaDataSize,
    return newDataStart;
    }
 
+uint8_t *
+TR_JITaaSRelocationRuntime::copyDataToCodeCache(const void *startAddress, size_t totalSize, TR_J9VMBase *fe)
+   {
+   TR::CompilationInfoPerThreadBase *compInfoPT = fe->_compInfoPT;
+   int32_t numReserved;
+   TR::CodeCache *codeCache = NULL;
+   TR::CodeCacheManager *manager = TR::CodeCacheManager::instance();
+   codeCache = manager->reserveCodeCache(false, totalSize, compInfoPT->getCompThreadId(), &numReserved);
+   if (!codeCache)
+      return NULL;
+
+   if (compInfoPT->getCompThreadId() >= 0 && fe->getCompilationShouldBeInterruptedFlag())
+      {
+      codeCache->unreserve();
+      return NULL;
+      }
+
+   uint8_t *coldCodeStart = NULL;
+   manager->allocateCodeMemory(0, totalSize, &codeCache, &coldCodeStart, false, false);
+   if (coldCodeStart)
+      {
+      memcpy(coldCodeStart, startAddress, totalSize);
```

Comment:
We have to unreserve even when the allocation was successful, because the next thunk will reserve again.
Also, let's put an assert that when we enter this method, no code cache is reserved.

Reply:
Done.